src/App.js

```diff
@@ -111,47 +117,111 @@ class App extends Component {
     this.setState({ isAuthenticating: false });
   }
 
   initialFetch = async (username) => {
-    try {
-      const user = await API.get("pareto", `/users/${username}`);
-      if (user.length > 0) {
-        this.props.getUser(user[0]);
-        this.setState({ user: user[0] });
-        if (user[0].defaultLanguage) {
+    let newState = {};
+    const path = this.props.location.pathname;
+
+    // Set up variables to enable fetching only the data needed for your current app view
+    const [context, training, arena] = [
+      path.includes("context-builder"),
+      path.includes("training"),
+      !path.includes("context-builder") && !path.includes("training"),
+    ];
+    const firstFetch = [];
+    const secondFetch = [];
+
+    const socketOptions = { updateState: false };
+
+    const user = await fetchUser(username);
+    if (user.length > 0) {
+      const currentUser = user[0];
+      try {
+        this.props.getUser(currentUser);
+        newState.user = currentUser;
+        if (currentUser.defaultLanguage) {
           const language = availableLanguages.find(
-            (x) => x.code === user[0].defaultLanguage
+            (x) => x.code === currentUser.defaultLanguage
           );
-          this.setState({
-            chosenLanguage: language || "en",
-          });
-          I18n.setLanguage(user[0].defaultLanguage);
+          newState.chosenLanguage = language || "en";
+          I18n.setLanguage(currentUser.defaultLanguage);
         }
-
-        // if (user[0].learningPurchase === true || user[0].instructor === true) {
-        await this.fetchStarterKitExperience(user[0].id);
-        await this.fetchStarterKitSanity();
-        // }
-
-        await this.fetchCoaches(user[0].id);
-        if (user[0].instructor === true) {
-          await this.fetchCoachingRoster(user[0].id);
+        socketOptions.user = currentUser;
+
+        // Sort fetching functions according to whether they should happen before or after the loading overlay goes away
+        if (context) {
+          firstFetch.push(fetchSanitySchemas);
+          firstFetch.push(fetchStarterKitSanity);
+        } else {
+          secondFetch.push(fetchSanitySchemas);
+          secondFetch.push(fetchStarterKitSanity);
+        }
+        if (training) {
+          firstFetch.push(() => fetchStarterKitExperience(currentUser.id));
+        } else {
+          secondFetch.push(() => fetchStarterKitExperience(currentUser.id));
+        }
+        if (arena) {
+          firstFetch.push(() => this.connectSocketToSprint(socketOptions));
+        } else {
+          secondFetch.push(() => this.connectSocketToSprint(socketOptions));
+        }
+        if (currentUser.instructor) {
+          firstFetch.push(() => fetchCoachingRoster(currentUser.id));
+        }
+        firstFetch.push(() => fetchCoaches(currentUser.id));
+
+        // Fetch first set of data
+        const results = await Promise.all([...firstFetch.map((x) => x())]);
+
+        results
```

Comment:
This is an interesting pattern, haven't seen this before! Works great for sorting the network request responses though.

Reply:
This is why I love JavaScript so much - there's always a new way to say something! 

I think overall this approach is more newcomer friendly than the way I originally had it. If you look back at earlier commits, I was creating the array of promises on the fly inside Promise.all, with logical conjunctions, and then filtering out the ones that returned false.